### PR TITLE
Rebase 'Add generic rumble gain to input settings' by davidgfnet

### PR DESCRIPTION
**Pull request by DisasterMo**
## Description

This is a rebase of #11702 + a bug fix by @iancharnas. Has been verified as working with xinput & udev drivers.

## Related Pull Requests

#11702

## Reviewers
@davidgfnet 
@jdgleaver

### CODE REVIEW

**Comment by davidgfnet** [comment]
Can you add @jdgleaver ?

**Comment by jdgleaver** [comment]
@DisasterMo Oh, I didn't realise that @davidgfnet had already added the input driver callbacks! Very nice indeed :)

There are just a few little issues here, two trivial and one a bit more significant:

- There is an odd mishmash of data types here: the base gain config value is `unsigned`, the input driver `set_rumble_gain()` function accepts a `float` and `frontend_driver_set_rumble_gain()` accepts an `int`. These should be standardised (probably to an `unsigned` in all cases)

- `frontend_driver_set_rumble_gain()` is in the wrong place. In order to follow the general schema, there should be an `input_driver_set_rumble_gain()` function in `input_driver.h/.c`, plus an `input_set_rumble_gain()` wrapper in `retroarch.h/.c` (just like we have with `input_set_rumble_state()`)

- `sdl_dingux_joypad.c` is now broken! A callback has been assigned, but the `sdl_dingux_joypad_set_rumble_gain()` function does not exist! Clearly this must be fixed before merging ;)

I'm in the middle of something else right now, but if you can address the first two points then I will prepare a little patch to fix `sdl_dingux_joypad.c` a bit later....

**Comment by DisasterMo** [comment]
@jdgleaver Good points, as always. I did not pay attention to types at all xD
I _was_ wondering, however, why there was no `sdl_dingux_joypad_set_rumble_gain()` function.
Well, I've replaced it with a `NULL` for now, since I don't really know how anything there works. 
I hope my implementation of `input_driver_set_rumble_gain()` & the `input_set_rumble_gain()` wrapper are acceptable.

File: input/drivers_joypad/udev_joypad.c [hunk 1]
```diff
@@ -147,6 +151,37 @@ static int udev_open_joystick(const char *path)
...
+
+   if (pad->fd < 0)
+      return false;
+
+   if (pad->rumble_gain == gain)
+      return true;
+
+   /* Does not support > 100 gains */
+   if (gain > 100)
+      return false;
```
Comment (DisasterMo) [inline input/drivers_joypad/udev_joypad.c:167]: @davidgfnet I was wondering, is this correct? Checking the gain against the upper bound _after_ attempting to apply it?

**Comment by jdgleaver** [comment]
@DisasterMo Good job, this all looks fine now

> I was wondering, however, why there was no sdl_dingux_joypad_set_rumble_gain() function.
Well, I've replaced it with a NULL for now, since I don't really know how anything there works.

Before we merge this, I will give you a patch to apply on-top of the PR which updates the sdl_dingux driver (it's a pretty simple change, but I need to test on real hardware before signing it off!). I'll try to do this tomorrow.

File: input/drivers_joypad/udev_joypad.c (hunk 1 above)
Comment (jdgleaver) [reply input/drivers_joypad/udev_joypad.c:167]: I think that looks okay - it's still performing the check before accessing the hardware. Could combine the checks, though, I suppose - i.e.:

```c
   struct input_event ie;
   struct udev_joypad *pad = (struct udev_joypad*)&udev_pads[i];

   /* Does not support > 100 gains */
   if ((pad->fd < 0) ||
       (gain > 100))
      return false;

   if (pad->rumble_gain == gain)
      return true;
```

**Comment by jdgleaver** [comment]
@DisasterMo Okay, here's the patch: [rumble_gain.patch.zip](https://github.com/libretro/RetroArch/files/7129675/rumble_gain.patch.zip)

It's highly trivial - just implements the `sdl_dingux_joypad_set_rumble_gain()` function, and adds a `break;` that was missing from the switch statement in `menu_cbs_sublabel.c`

Tested on OpenDingux, this works very well - the approach taken in this PR greatly improves the user experience!

There's one last little thing to address: the `(Restart Required)` in `Vibration Strength (Restart Required)` is no longer valid - all updates now happen in realtime. If you can add the patch and remove the bracketed text, this will be ready to merge :)

**Comment by DisasterMo** [comment]
Thank you very much, @jdgleaver !

I've added all changes.

Thank you as well, @davidgfnet , for providing the idea & original code!

**Comment by davidgfnet** [comment]
Thanks to you for doing this, really!

**Review by jdgleaver — APPROVED** [review]

**Comment by jdgleaver** [comment]
Thanks to you both! This is a huge QoL improvement :)

**Pull request merged**